controllers/metal3.io/baremetalhost_controller.go

```diff
@@ -1153,6 +1179,227 @@ func saveHostProvisioningSettings(host *metal3v1alpha1.BareMetalHost) (dirty boo
+	if hfs.Status.FirmwareSchema != nil {
+		if err = r.Get(context.TODO(), client.ObjectKey{
+			Namespace: hfs.Status.FirmwareSchema.Namespace,
+			Name:      hfs.Status.FirmwareSchema.Name}, schema); err != nil {
+
+			info.log.Error(err, "failed to get schema for hostFirmwareSettings")
+			return nil, err
+		}
+	} else {
+		schema = nil
+	}
+
+	// Validate the Spec setting using schema, if available
+	if err = ValidateHostFirmwareSettings(hfs, schema); err != nil {
+		return nil, err
+	}
+
+	return hfs, nil
+}
+
+// Get the firmware settings from the provisioner and create hostFirmwareSettings and FirmwareSchema resources
+func (r *BareMetalHostReconciler) storeHostFirmwareSettings(prov provisioner.Provisioner, info *reconcileInfo) (err error) {
+
+	info.log.Info("storeHostFirmwareSettings")
+
+	// First check if there is a hostFirmwareSettings resource for this host
+	hfs := &metal3v1alpha1.HostFirmwareSettings{}
+	if err = r.Get(context.TODO(), info.request.NamespacedName, hfs); err == nil {
+		info.log.Info("found existing hostFirmwareSettings resource")
+	} else if !k8serrors.IsNotFound(err) {
+		// Error reading the object
+		return errors.Wrap(err, "could not load host firmware settings")
+	} else {
+		// Create a new resource
+		if err = r.newHostFirmwareSettings(info, hfs); err != nil {
+			return errors.Wrap(err, "failure calling newHostFirmwareSettings")
+		}
+	}
+
+	// Get the current settings and schema
+	currentSettings, schema, err := prov.GetFirmwareSettings(true)
+	if err != nil {
+		info.log.Info("could not get firmware settings from provisioner")
+		return err
+	}
+
+	if schema != nil && len(schema) > 0 {
+
+		firmwareSchema, err := r.getOrCreateFirmwareSchema(info, hfs, schema)
+		if err != nil {
+			return errors.Wrap(err, "could not get/create firmware schema in cluster")
+		}
+
+		// Set hostFirmwareSetting to use this schema
+		hfs.Status.FirmwareSchema = &metal3v1alpha1.SchemaReference{
+			Namespace: firmwareSchema.ObjectMeta.Namespace,
+			Name:      firmwareSchema.ObjectMeta.Name}
+
+		// copy schema reference to Annotation so it can be restored after pivoting
+		hfs.Annotations[metal3v1alpha1.FirmwareSchemaNamespaceAnnotation] = hfs.Status.FirmwareSchema.Namespace
+		hfs.Annotations[metal3v1alpha1.FirmwareSchemaNameAnnotation] = hfs.Status.FirmwareSchema.Name
+
+		if err = r.updateHostFirmwareSettingsFromCurrent(currentSettings, hfs, firmwareSchema); err != nil {
+			return errors.Wrap(err, "could not update hostFirmwareSettings")
+		}
+	} else {
+		if err = r.updateHostFirmwareSettingsFromCurrent(currentSettings, hfs, nil); err != nil {
+			return errors.Wrap(err, "could not update hostFirmwareSettings")
+		}
+		info.log.Info("updated hostFirmwareSettings settings without schema")
+	}
+
+	return nil
+}
+
+// Create a HostFirmwareSettings resource
+func (r *BareMetalHostReconciler) newHostFirmwareSettings(info *reconcileInfo, hfs *metal3v1alpha1.HostFirmwareSettings) (err error) {
+
```

Comment:
I would have expected to find the `hfs` creation directly within the factory method (otherwise it's more an `update` rather than a `new` one)